File: python/purehtml/transformers/types/remove_last_path_section_transformer.py
from abc import ABC
from typing import List
from urllib.parse import urlsplit
from urllib.parse import urlunsplit

from purehtml.transformers.transform_params import TransformParams
from purehtml.transformers.transformer import Transformer


class RemoveLastPathSectionTransformer(Transformer, ABC):
    def __init__(self, args: List[str] = None):
        """
        Initialize the transformer with optional arguments.
        :param args: Arguments to configure the transformer (unused in this case).
        """
        super().__init__()
        self.args = args if args is not None else []

    def set_args(self, args: List[str]):
        """
        Set the list of attributes to retrieve.
        :param args: A list of attribute names.
        """
        self.args = args

    def get_name(self) -> str:
        """
        Return the name of the transformer.
        :return: The name of the transformer.
        """
        return "removeLastPathSection"

    def transform(self, params: TransformParams) -> object:
        """
        Remove the last section of the URL's path.
        :param params: The parameters that contain the value (a URL) to be transformed.
        :return: The URL with its last path section removed.
        """
        val = params.get_val()

        if val is None:
            return None

        if not isinstance(val, str):
            raise TypeError(
                f"removeLastPathSection: expects string. Got {val!r}"
            )

        parts = urlsplit(val)
        sections = parts.path.split("/")
        sections.pop()

        path = "/".join(sections)
        if path == "":
            path = "/"

        return urlunsplit((parts.scheme, parts.netloc, path, parts.query, parts.fragment))

File: java/src/main/java/com/purescraps/purehtml/transformers/types/RemoveLastPathSectionTransformer.java
package com.purescraps.purehtml.transformers.types;

import java.net.URI;
import java.net.URISyntaxException;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

import com.purescraps.purehtml.transformers.TransformParams;
import com.purescraps.purehtml.transformers.Transformer;

public class RemoveLastPathSectionTransformer extends Transformer {
    @SuppressWarnings("unused")
    private List<String> args;

    @SuppressWarnings("unused")
    public RemoveLastPathSectionTransformer(List<String> args) {
        this.args = args;
    }

    @SuppressWarnings("unused")
    public RemoveLastPathSectionTransformer() {
        this.args = null;
    }

    public void setArgs(List<String> args) {
        this.args = args;
    }

    // Static method to return the name of the transformer
    public String getName() {
        return "removeLastPathSection";
    }

    // The transform method that removes the last section of the URL's path
    @Override
    public Object transform(TransformParams params) {
        Object val = params.getVal();

        if (val == null) {
            return null;
        }

        if (!(val instanceof String)) {
            throw new IllegalArgumentException("removeLastPathSection: expects string. Got " + val);
        }

        try {
            URI uri = new URI((String) val);
            String path = uri.getPath() == null ? "" : uri.getPath();

            List<String> sections = new ArrayList<>(Arrays.asList(path.split("/", -1)));
            if (!sections.isEmpty()) {
                sections.remove(sections.size() - 1);
            }

            String newPath = String.join("/", sections);
            if (newPath.isEmpty()) {
                newPath = "/";
            }

            return new URI(uri.getScheme(), uri.getAuthority(), newPath, uri.getQuery(), uri.getFragment())
                    .toString();
        } catch (URISyntaxException e) {
            throw new IllegalArgumentException("removeLastPathSection: invalid URL: " + val, e);
        }
    }
}
